File: blue-core/src/main/java/blue/soundObject/Instance.java
package blue.soundObject;

import blue.score.ScoreObjectEvent;
import blue.*;
import blue.noteProcessor.NoteProcessorChain;
import blue.noteProcessor.NoteProcessorException;
import blue.utility.ScoreUtilities;
import electric.xml.Element;
import java.util.Map;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 *
 * @author steven yi
 * @version 1.0
 */
public class Instance extends AbstractSoundObject {

//    private static BarRenderer renderer = new LetterRenderer("I");
    private SoundObject sObj;

    private NoteProcessorChain npc = new NoteProcessorChain();

    private TimeBehavior timeBehavior;

    double repeatPoint = -1.0f;

    /*
     * cache library id when loading up the SoundObjectLibrary, to be resolved
     * in second pass
     * FIXME: THIS NEEDS TO BE REPLACED!
     */
    int soundObjectLibraryId = -1;

    public Instance(SoundObject sObj) {
        this.sObj = sObj;
        setName(sObj.getName());
        this.setBackgroundColor(sObj.getBackgroundColor());
        timeBehavior = TimeBehavior.SCALE;
    }

    public Instance() {
        this.name = "Instance: ";
    }

    /**
     * Copy Constructor NOTE: intentionally keeps same SoundObject reference in
     * copy as original
     *
     * @param instance
     */
    public Instance(Instance instance) {
        super(instance);
        sObj = instance.sObj;
        npc = new NoteProcessorChain(instance.npc);
        backgroundColor = instance.getBackgroundColor();
        timeBehavior = instance.getTimeBehavior();
        repeatPoint = instance.repeatPoint;
    }

    public void processNotes(NoteList nl) throws SoundObjectException {

        ScoreUtilities.normalizeNoteList(nl);

        try {
           nl =  ScoreUtilities.applyNoteProcessorChain(nl, this.npc);
        } catch (NoteProcessorException e) {
            throw new SoundObjectException(this, e);
        }

        ScoreUtilities.applyTimeBehavior(nl, this.getTimeBehavior(), this
                .getSubjectiveDuration(), this.getRepeatPoint());
        ScoreUtilities.setScoreStart(nl, startTime);

    }

    @Override
    public double getObjectiveDuration() {
        return sObj.getSubjectiveDuration();
    }

    @Override
    public void setNoteProcessorChain(NoteProcessorChain npc) {
        this.npc = npc;
    }

    @Override
    public NoteProcessorChain getNoteProcessorChain() {
        return npc;
    }

    @Override
    public TimeBehavior getTimeBehavior() {
        return this.timeBehavior;
    }

    @Override
    public void setTimeBehavior(TimeBehavior timeBehavior) {
        this.timeBehavior = timeBehavior;
    }

    @Override
    public double getRepeatPoint() {
        return this.repeatPoint;
    }

    @Override
    public void setRepeatPoint(double repeatPoint) {
        this.repeatPoint = repeatPoint;

        ScoreObjectEvent event = new ScoreObjectEvent(this,
                ScoreObjectEvent.REPEAT_POINT);

        fireScoreObjectEvent(event);
    }

    /*
     * (non-Javadoc)
     *
     * @see blue.soundObject.SoundObject#loadFromXML(electric.xml.Element)
     */
    public static SoundObject loadFromXML(Element data,
            Map<String, Object> objRefMap) throws Exception {
        Instance instance = new Instance();

        SoundObjectUtilities.initBasicFromXML(data, instance);

        String id = data.getElement("soundObjectReference")
                .getAttributeValue("soundObjectLibraryID");

        if ("null".equals(id)) {
            throw new Exception("ERROR: SoundObject Instance found pointing to an library item that no longer exists");
        }

        Object sObj = objRefMap.get(id);
        if (sObj != null) {
            instance.setSoundObject((SoundObject) sObj);
        } else {
            throw new Exception("Could not find SoundObject pointed to from Instance with ID: " + id);
        }

        return instance;

    }

    /*
     * (non-Javadoc)
     *
     * @see blue.soundObject.SoundObject#saveAsXML()
     */
    @Override
    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element retVal = SoundObjectUtilities.getBasicXML(this);

        retVal.addElement("soundObjectReference").setAttribute(
                "soundObjectLibraryID",
                objRefMap.get(this
                        .getSoundObject()));

        return retVal;
    }

    /**
     * @return Returns the sObj.
     */
    public SoundObject getSoundObject() {
        return sObj;
    }

    /**
     * @param obj The sObj to set.
     */
    public void setSoundObject(SoundObject obj) {
        sObj = obj;
    }

    @Override
    public NoteList generateForCSD(CompileData compileData, double startTime, double endTime)
            throws SoundObjectException {
        NoteList nl = sObj.generateForCSD(compileData, startTime, endTime);
        processNotes(nl);

        return nl;
    }

    @Override
    public Instance deepCopy() {
        return new Instance(this);
    }
}

File: packages/blue-data/src/sound-objects/instance.ts
/**
 * Instance — references a SoundObject from the SoundObjectLibrary.
 * Mirrors the Java Instance class.
 *
 * Delegates note generation to the referenced SoundObject, then applies
 * its own note processor chain and time behavior.
 */
import { AbstractSoundObject } from './abstract-sound-object';
import { NoteList } from './note-list';
import { TimeContext } from '../time/time-context';
import { CompileData } from '../compile-data';
import { Element } from '../serialization/xml-reader';
import { ObjRefSaveMap, ObjRefLoadMap } from '../serialization/obj-ref-map';
import { SoundObject } from './sound-object';
import { initBasicFromXML, getBasicXML } from './sound-object-utilities';
import { setScoreStart } from '../utilities/score';

export class Instance extends AbstractSoundObject {
  private _soundObject: SoundObject | null = null;
  private _libraryId = '';

  constructor(other?: Instance) {
    super();
    this.setName('Instance: ');
    if (other) {
      this.copyFrom(other);
      this._libraryId = other._libraryId;
      // Note: _soundObject reference is shared (same as Java copy constructor)
      this._soundObject = other._soundObject;
    }
  }

  getSoundObject(): SoundObject | null { return this._soundObject; }
  setSoundObject(sObj: SoundObject): void {
    this._soundObject = sObj;
    this.setName(sObj.getName());
    this.setBackgroundColor(sObj.getBackgroundColor());
  }

  getLibraryId(): string { return this._libraryId; }
  setLibraryId(id: string): void { this._libraryId = id; }


  override generateForCSD(
    context: TimeContext,
    compileData: CompileData,
    startTime: number,
    endTime: number,
  ): NoteList {
    if (!this._soundObject) {
      return new NoteList();
    }

    const nl = this._soundObject.generateForCSD(context, compileData, startTime, endTime);

    // Apply note processor chain
    const npc = this.getNoteProcessorChain();
    npc.apply(nl);

    // Apply time behavior
    const duration = this._subjectiveDuration.toBeats(context);
    const rpBeats = this._repeatPoint ? this._repeatPoint.toBeats(context) : -1;
    // Note: full time behavior application needs ScoreUtilities — simplified for Phase 11
    setScoreStart(nl, this._startTime.toBeats(context));

    return nl;
  }

  async generateForCSDAsync(
    context: TimeContext,
    compileData: CompileData,
    startTime: number,
    endTime: number,
  ): Promise<NoteList> {
    if (!this._soundObject) {
      return new NoteList();
    }

    const nl = this._soundObject.generateForCSDAsync
      ? await this._soundObject.generateForCSDAsync(context, compileData, startTime, endTime)
      : this._soundObject.generateForCSD(context, compileData, startTime, endTime);

    const npc = this.getNoteProcessorChain();
    npc.apply(nl);
    setScoreStart(nl, this._startTime.toBeats(context));

    return nl;
  }

  override saveAsXML(objRefMap?: ObjRefSaveMap): Element {
    const elem = getBasicXML(this, 'blue.soundObject.Instance');

    const refElem = elem.addElement('soundObjectReference');
    if (this._soundObject && objRefMap) {
      refElem.setAttribute('soundObjectLibraryID', objRefMap.getId(this._soundObject));
    } else {
      refElem.setAttribute('soundObjectLibraryID', this._libraryId || 'null');
    }

    return elem;
  }

  static loadFromXML(data: Element, objRefMap?: ObjRefLoadMap): Instance {
    const obj = new Instance();
    initBasicFromXML(obj, data);

    const refNode = data.getElement('soundObjectReference');
    if (refNode) {
      const id = refNode.getAttribute('soundObjectLibraryID') ?? 'null';
      if (id === 'null' || !id) {
        // Unresolved reference — store as library id for later resolution
        obj._libraryId = '';
      } else if (objRefMap && objRefMap.has(id)) {
        obj._soundObject = objRefMap.get(id) as SoundObject;
        obj._libraryId = id;
      } else {
        obj._libraryId = id;
        // Will be resolved in second pass by caller
      }
    }

    return obj;
  }

  override deepCopy(): SoundObject {
    const copy = new Instance(this);
    return copy;
  }
}
